lsp/nls/src/background.rs

```diff
@@ -0,0 +1,347 @@
+
+const EVAL_TIMEOUT: Duration = Duration::from_secs(1);
+
+#[derive(Debug, Serialize, Deserialize)]
+enum Command {
+    UpdateFile { uri: Url, text: String },
+    EvalFile { uri: Url },
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Diagnostics {
+    pub path: PathBuf,
+    pub diagnostics: Vec<SerializableDiagnostic>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Response {
+    Diagnostics(Diagnostics),
+    /// The background worker sends back one of these when it's about to start
+    /// an eval job. That way, if it becomes unresponsive we know which file is
+    /// the culprit.
+    Starting {
+        uri: Url,
+    },
+}
+
+pub struct BackgroundJobs {
+    receiver: Receiver<Diagnostics>,
+    sender: Sender<Command>,
+}
+
+// The entry point of the background worker. If it fails to bootstrap the connection,
+// panic immediately (it's a subprocess anyway).
+pub fn worker_main(main_server: String) {
+    let oneshot_tx = IpcSender::connect(main_server).unwrap();
+    let (cmd_tx, cmd_rx) = ipc_channel::ipc::channel().unwrap();
+    let (response_tx, response_rx) = ipc_channel::ipc::channel().unwrap();
+    oneshot_tx.send((cmd_tx, response_rx)).unwrap();
+
+    worker(cmd_rx, response_tx);
+}
+
+fn drain_ready<T: for<'de> Deserialize<'de> + Serialize>(rx: &IpcReceiver<T>, buf: &mut Vec<T>) {
+    while let Ok(x) = rx.try_recv() {
+        buf.push(x);
+    }
+}
+
+// Returning an Option, just to let us use `?` when channels disconnect.
+fn worker(cmd_rx: IpcReceiver<Command>, response_tx: IpcSender<Response>) -> Option<()> {
+    let mut evals = VecDeque::new();
+    let mut cmds = Vec::new();
+    let mut world = World::default();
+
+    drain_ready(&cmd_rx, &mut cmds);
+    loop {
+        for cmd in cmds.drain(..) {
+            // Process all the file updates first, even if it's out of order with the evals.
+            // (Is there any use case for wanting the eval before updating the contents?)
+            match cmd {
+                Command::UpdateFile { uri, text } => {
+                    // Failing to update a file's contents is bad, so we terminate (and restart)
+                    // the worker.
+                    world.update_file(uri, text).unwrap();
+                }
+                Command::EvalFile { uri } => {
+                    evals.push_back(uri);
+                }
+            }
+        }
+
+        // Deduplicate the evals back-to-front, and then evaluate front-to-back.
+        // This means that if we get requests for `foo.ncl` and `bar.ncl`, and
+        // then another request for `foo.ncl` before started working on the
+        // first one, then we'll throw away the first `foo.ncl` and evaluate
+        // `bar.ncl` first.
+        let mut seen = HashSet::new();
+        let mut dedup = VecDeque::new();
+        for path in evals.iter().rev() {
+            if seen.insert(path) {
+                dedup.push_front(path.clone());
```

Comment:
Nitpick: couldn't we just use a `Vec`, `push` and then iterate in reverse for dedup? As it seems we always push only on one side (front)